azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/extensions/AzureComputeImageExtension.java

```diff
@@ -0,0 +1,149 @@
+import static org.jclouds.compute.config.ComputeServiceProperties.TIMEOUT_IMAGE_AVAILABLE;
+import com.google.common.util.concurrent.UncheckedTimeoutException;
+import org.jclouds.util.Predicates2;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+
+
+public class AzureComputeImageExtension implements ImageExtension {
+   private final AzureComputeApi api;
+   private final Predicate<URI> imageAvailablePredicate;
+   private final ListeningExecutorService userExecutor;
+   private final String group;
+   private final VMImageToImage imageReferenceToImage;
+   public static final String CONTAINER_NAME = "vhdsnew";
+   public static final String CUSTOM_IMAGE_PREFIX = "#";
+
+   @Inject
+   AzureComputeImageExtension(AzureComputeApi api,
+                              @Named(TIMEOUT_IMAGE_AVAILABLE) Predicate<URI> imageAvailablePredicate,
+                              final AzureComputeServiceContextModule.AzureComputeConstants azureComputeConstants,
+                              @Named(Constants.PROPERTY_USER_THREADS) ListeningExecutorService userExecutor,
+                              VMImageToImage imageReferenceToImage) {
+      this.userExecutor = userExecutor;
+      this.group = azureComputeConstants.azureResourceGroup();
+      this.imageReferenceToImage = imageReferenceToImage;
+      this.api = api;
+      this.imageAvailablePredicate = imageAvailablePredicate;
+   }
+
+   @Override
+   public ImageTemplate buildImageTemplateFromNode(String name, String id) {
+      String imageName = name.toLowerCase();
+      return new ImageTemplateBuilder.CloneImageTemplateBuilder().nodeId(id).name(imageName).build();
+   }
+
+   @Override
+   public ListenableFuture<Image> createImage(ImageTemplate template) {
+      final CloneImageTemplate cloneTemplate = (CloneImageTemplate) template;
+      final String id = cloneTemplate.getSourceNodeId();
+      final String storageAccountName = id.replaceAll("[^A-Za-z0-9 ]", "") + "stor";
+
+      // VM needs to be stopped before it can be generalized
+      String status = "";
+      api.getVirtualMachineApi(group).stop(id);
+      //Poll until resource is ready to be used
+      boolean jobDone = Predicates2.retry(new Predicate<String>() {
+         @Override
+         public boolean apply(String name) {
+            String status = "";
+            List<VirtualMachineInstance.VirtualMachineStatus> statuses = api.getVirtualMachineApi(group).getInstanceDetails(name).statuses();
+            for (int c = 0; c < statuses.size(); c++) {
+               if (statuses.get(c).code().substring(0, 10).equals("PowerState")) {
+                  status = statuses.get(c).displayStatus();
+                  break;
+               }
+            }
+            if (status.equals("VM stopped")) {
+               return true;
+            }
+            return false;
+         }
+      }, 60 * 4 * 1000).apply(id);
```

Comment:
Refactor this in a reusable and injectable predicate the same way than has been done with the image available one. Also, the timeout to be used should be the [TIMEOUT_NODE_SUSPENDED](https://github.com/jclouds/jclouds/blob/master/compute/src/main/java/org/jclouds/compute/reference/ComputeServiceConstants.java#L103-L105).